cpp/src/arrow/util/ubsan.h

```diff
@@ -60,6 +60,14 @@ inline typename std::enable_if<std::is_trivial<T>::value, T>::type SafeLoadAs(
   return ret;
 }
 
+template <typename T>
+inline typename std::enable_if<std::is_trivial<T>::value, T>::type SafeLoadAs(
+    const uint8_t* unaligned, uint32_t length) {
+  typename std::remove_const<T>::type ret;
+  std::memcpy(&ret, unaligned, length);
```

Comment:
Looks there might be some garbage bytes in the `ret` as it's only partially written?

Reply:
SafeLoadAs with length is only called by `helper_tail`. The garbage bytes are filtered out by a mask. 
In the original implementation, the garbage bytes are also copied, if the `unaligned` doesn't have enough length.

Reply:
Might be a good idea to explain the difference between with/without length in a comment here, in case someone tries to reuse this function later.

Reply:
I agree with @cyb70289 that we should zero-initialize the result.

Reply:
The result `ret` is initialized to 0, and comments are added.